kyo-combinators/shared/src/main/scala/kyo/combinators.scala

```diff
@@ -0,0 +1,429 @@
+        f: Flat[A]
+    ): [A1 >: A, S1] => PartialFunction[E1, A1 < S1] => A1 < (S & S1 & Aborts[E]) =
+        [A1 >: A, S1] =>
+            (fn: PartialFunction[E1, A1 < S1]) =>
+                Aborts.run[E1](effect).map {
+                    case Left(e1) if fn.isDefinedAt(e1) => fn(e1)
+                    case Left(e1)                       => Aborts.fail[E1](e1)
+                    case Right(a)                       => a
+                    // Need asInstanceOf because compiler doesn't know ha.Remainder & Aborts[E1]
+                    // is the same as Aborts[E]
+                }.asInstanceOf[A1 < (S & S1 & Aborts[E])]
+
+    def swapAborts(
+        using
+        cte: ClassTag[E],
+        cta: ClassTag[A],
+        fl: Flat[A]
+    ): E < (S & Aborts[A]) =
+        val handled: Either[E, A] < S = Aborts.run[E](effect)
+        handled.map((v: Either[E, A]) => Aborts.get(v.swap))
+    end swapAborts
+
+    def swapSomeAborts[E1: ClassTag](
+        using
+        ha: Aborts.HasAborts[E1, E],
+        cte: ClassTag[E],
+        cta: ClassTag[A],
+        f: Flat[A]
+    ): E1 < (S & ha.Remainder & Aborts[A]) =
+        val handled = Aborts.run[E1](effect)
+        handled.map((v: Either[E1, A]) => Aborts.get(v.swap))
+    end swapSomeAborts
+
+    def implicitThrowable(
+        using
+        f: Flat[A],
+        ha: Aborts.HasAborts[Throwable, E]
+    ): A < (S & ha.Remainder) =
+        Aborts.run[Throwable](effect).map {
+            case Right(a) => a
+            case Left(e)  => throw e
+        }
+end extension
+
+extension [A, S](effect: A < (S & Options))
+    def handleOptions(
+        using Flat[A]
+    ): Option[A] < S = Options.run(effect)
+
+    def catchOptions[A1 >: A, S1](orElse: => A1 < S1)(
+        using Flat[A]
+    ): A1 < (S & S1) =
+        Options.run(effect).map {
+            case None    => orElse
+            case Some(a) => a
+        }
+
+    def swapOptionsAs[A1, S1](value: => A1 < S1)(
+        using Flat[A]
+    ): A1 < (S & S1 & Options) =
+        Options.run(effect).map {
+            case None    => value
+            case Some(a) => Options.empty
+        }
+
+    def swapOptions(using Flat[A]): Unit < (S & Options) =
+        swapOptionsAs(())
+
+    def optionsToAborts[E, S1](failure: => E < S1)(
+        using Flat[A]
+    ): A < (S & S1 & Aborts[E]) =
+        Options.run(effect).map {
+            case None    => Kyo.fail(failure)
+            case Some(a) => a
+        }
+
+    def optionsToThrowable(using Flat[A]): A < (S & Aborts[Throwable]) =
+        effect.optionsToAborts[Throwable, Any](new NoSuchElementException("None.get"))
+
+    def optionsToUnit(using Flat[A]): A < (S & Aborts[Unit]) =
+        effect.optionsToAborts(())
+
+    def optionsToChoices(using Flat[A]): A < (S & Choices) =
+        Options.run(effect).map(aOpt => Choices.get(aOpt.toSeq))
+end extension
+
+extension [A, S, E](effect: A < (S & Envs[E]))
+    def provide[E1, S1, SR](dependency: E1 < S1)(
```

Comment:
Unless you are already using these methods, we will need to implement a stronger version of Provides for the layer implementation. We can later just use this extension as syntax over the `Layer.provide`

Reply:
Yeah provides should definitely be for layers. That's not finished though, yet, is it?